tensorboard/notebook.py

```diff
@@ -0,0 +1,336 @@
+        display_handle=handle,
+    )
+
+  elif isinstance(start_result, manager.StartReused):
+    template = (
+        "Reusing TensorBoard on port {port} (pid {pid}), started {delta} ago. "
+        "(Use '!kill {pid}' to kill it.)"
+    )
+    message = template.format(
+        port=start_result.info.port,
+        pid=start_result.info.pid,
+        delta=_time_delta_from_info(start_result.info),
+    )
+    print_or_update(message)
+    _display(
+        port=start_result.info.port,
+        print_message=False,
+        display_handle=None,
+    )
+
+  elif isinstance(start_result, manager.StartFailed):
+    def format_stream(name, value):
+      if value == "":
+        return ""
+      elif value is None:
+        return "\n<could not read %s>" % name
+      else:
+        return "\nContents of %s:\n%s" % (name, value.strip())
+    message = (
+        "ERROR: Failed to launch TensorBoard (exited with %d).%s%s" %
+        (
+            start_result.exit_code,
+            format_stream("stderr", start_result.stderr),
+            format_stream("stdout", start_result.stdout),
+        )
+    )
+    print_or_update(message)
+
+  elif isinstance(start_result, manager.StartTimedOut):
+    message = (
+        "ERROR: Timed out waiting for TensorBoard to start. "
+        "It may still be running as pid %d."
+        % start_result.pid
+    )
+    print_or_update(message)
+
+  else:
+    raise TypeError(
+        "Unexpected result from `manager.start`: %r.\n"
+        "This is a TensorBoard bug; please report it."
+        % start_result
+    )
+
+
+def _time_delta_from_info(info):
+  """Format the elapsed time for the given TensorboardInfo.
+
+  Args:
+    info: A TensorboardInfo value.
+
+  Returns:
+    A human-readable string describing the time since the server
+    described by `info` started: e.g., "2 days, 0:48:58".
+  """
+  now = datetime.datetime.now()
+  then = info.start_time
+  return str(now.replace(microsecond=0) - then.replace(microsecond=0))
+
+
+def _data_source_from_info(info):
+  """Format the data location for the given TensorboardInfo.
+
+  Args:
+    info: A TensorboardInfo value.
+
+  Returns:
+    A human-readable string describing the logdir or database connection
+    used by the server: e.g., "logdir /tmp/logs".
+  """
+  if info.db:
+    return "db %s" % info.db
+  else:
+    return "logdir %s" % info.logdir
+
+
+def display(port=None, height=None):
+  """Display a TensorBoard instance already running on this machine.
+
+  Args:
+    port: The port on which the TensorBoard server is listening, as an
+      `int`, or `None` to automatically select the most recently
+      launched TensorBoard.
+    height: The height of the frame into which to render the TensorBoard
+      UI, as an `int` number of pixels, or `None` to use a default value.
+  """
+  _display(port=port, height=height, print_message=True, display_handle=None)
+
+
+def _display(port=None, height=None, print_message=False, display_handle=None):
```

Comment:
Please comment what the default height is.

Reply:
Happy to, but I’m curious why? I’ve always considered “cosmetic” default
values like this one (as opposed to semantically important ones, like
`SortOrder.ASCENDING`) as implementation details that shouldn’t be
documented.

Reply:
Because height doesn't need to be specified and the caller may want to know what will happen if they don't specify it.

Reply:
FWIW, it seems to me that they _are_ told what will happen if they don’t
specify it: “…or `None` to use a default value”. But, changed anyway. :-)